clickhouse/columns/string.cpp

```diff
@@ -42,104 +70,187 @@ void ColumnFixedString::Append(ColumnRef column) {
 }
 
 bool ColumnFixedString::Load(CodedInputStream* input, size_t rows) {
-    data_.reserve(data_.size() + rows);
-
-    for (size_t i = 0; i < rows; ++i) {
-        std::string s;
-        s.resize(string_size_);
-
-        if (!WireFormat::ReadBytes(input, &s[0], s.size())) {
-            return false;
-        }
-
-        data_.push_back(std::move(s));
+    data_.resize(string_size_ * rows);
+    if (!WireFormat::ReadBytes(input, &data_[0], data_.size())) {
+        return false;
     }
 
     return true;
 }
 
 void ColumnFixedString::Save(CodedOutputStream* output) {
-    for (size_t i = 0; i < data_.size(); ++i) {
-        WireFormat::WriteBytes(output, data_[i].data(), string_size_);
-    }
+    WireFormat::WriteBytes(output, data_.data(), data_.size());
 }
 
 size_t ColumnFixedString::Size() const {
-    return data_.size();
+    return data_.size() / string_size_;
 }
 
 ColumnRef ColumnFixedString::Slice(size_t begin, size_t len) {
     auto result = std::make_shared<ColumnFixedString>(string_size_);
 
-    if (begin < data_.size()) {
-        result->data_ = SliceVector(data_, begin, len);
+    if (begin < Size()) {
+        const auto b = begin * string_size_;
+        const auto l = len * string_size_;
+        result->data_ = data_.substr(b, std::min(data_.size() - b, l));
     }
 
-    return result;
+    return std::move(result);
 }
 
+struct ColumnString::Block
+{
+    using CharT = typename std::string::value_type;
+
+    explicit Block(size_t starting_capacity)
+        : size(0),
+          capacity(starting_capacity),
+          data_(new CharT[capacity])
+    {}
+
+    inline auto GetAvailble() const
+    {
+        return capacity - size;
+    }
+
+    std::string_view AppendUnsafe(std::string_view str)
+    {
+        const auto pos = &data_[size];
+
+        memcpy(pos, str.data(), str.size());
+        size += str.size();
+
+        return std::string_view(pos, str.size());
+    }
+
+    auto GetCurrentWritePos()
+    {
+        return &data_[size];
+    }
+
+    std::string_view ConsumeTailAsStringViewUnsafe(size_t len)
+    {
+        const auto start = &data_[size];
+        size += len;
+        return std::string_view(start, len);
+    }
+
+    size_t size;
+    const size_t capacity;
+    std::unique_ptr<CharT[]> data_;
+};
 
 ColumnString::ColumnString()
     : Column(Type::CreateString())
 {
 }
 
-ColumnString::ColumnString(const std::vector<std::string>& data)
+ColumnString::ColumnString(const std::vector<std::string> & data)
     : Column(Type::CreateString())
-    , data_(data)
 {
+    items_.reserve(data.size());
+    blocks_.emplace_back(ComputeTotalSize(data));
+
+    for (const auto & s : data)
+    {
+        AppendUnsafe(s);
+    }
 }
 
-void ColumnString::Append(const std::string& str) {
-    data_.push_back(str);
+ColumnString::~ColumnString()
```

Comment:
Nope, just a trick to explicitly instantiate d-tor (and hence d-tor of `std::unique_ptr<Block[]>`) in the context where Block is defined. Otherwise said  `unique_ptr` fails to compile.

Reply:
Hm, strange. Doesn't `~ColumnString() = default` help?

Reply:
IIRC that does not prescribe where the d-tor is instantiated, which might be anywhere, like in every other translation unit.